prow/git/v2/interactor.go

```diff
@@ -0,0 +1,279 @@
+	out, err := i.executor.Run("rev-parse", commitlike)
+	if err != nil {
+		return "", fmt.Errorf("error parsing %q: %v %v", commitlike, err, string(out))
+	}
+	return string(out), nil
+}
+
+// BranchExists returns true if branch exists in heads.
+func (i *interactor) BranchExists(branch string) bool {
+	i.logger.Infof("Checking if branch %q exists", branch)
+	_, err := i.executor.Run("ls-remote", "--exit-code", "--heads", "origin", branch)
+	return err == nil
+}
+
+// CheckoutNewBranch creates a new branch and checks it out.
+func (i *interactor) CheckoutNewBranch(branch string) error {
+	i.logger.Infof("Checking out new branch %q", branch)
+	if out, err := i.executor.Run("checkout", "-b", branch); err != nil {
+		return fmt.Errorf("error checking out new branch %q: %v %v", branch, err, string(out))
+	}
+	return nil
+}
+
+// Merge attempts to merge commitlike into the current branch. It returns true
+// if the merge completes. It returns an error if the abort fails.
+func (i *interactor) Merge(commitlike string) (bool, error) {
+	return i.MergeWithStrategy(commitlike, "merge")
+}
+
+// MergeWithStrategy attempts to merge commitlike into the current branch given the merge strategy.
+// It returns true if the merge completes. It returns an error if the abort fails.
+func (i *interactor) MergeWithStrategy(commitlike, mergeStrategy string) (bool, error) {
+	i.logger.Infof("Merging %q using the %q strategy", commitlike, mergeStrategy)
+	mergeFlag := ""
+	switch mergeStrategy {
+	case "merge":
+		mergeFlag = "--no-ff"
+	case "squash":
+		mergeFlag = "--squash"
+	default:
+		return false, fmt.Errorf("merge strategy %q is not supported", mergeStrategy)
+	}
+	out, err := i.executor.Run("merge", mergeFlag, "--no-stat", "-m", "merge", commitlike)
+	if err == nil {
+		return true, nil
+	}
+	i.logger.WithError(err).Warnf("Error merging %q: %s", commitlike, string(out))
+	if out, err := i.executor.Run("merge", "--abort"); err != nil {
+		return false, fmt.Errorf("error aborting merge of %q: %v %v", commitlike, err, string(out))
+	}
+	return false, nil
+}
+
+// Only the `merge` and `squash` strategies are supported.
+func (i *interactor) MergeAndCheckout(baseSHA string, headSHAs []string, mergeStrategy string) error {
+	if baseSHA == "" {
+		return errors.New("baseSHA must be set")
+	}
+	if len(headSHAs) == 0 {
+		return errors.New("at least one headSHA must be provided")
+	}
+	if err := i.Checkout(baseSHA); err != nil {
+		return err
+	}
+	for _, headSHA := range headSHAs {
+		ok, err := i.MergeWithStrategy(headSHA, mergeStrategy)
+		if err != nil {
+			return err
+		} else if !ok {
+			return fmt.Errorf("failed to merge %q", headSHA)
+		}
+	}
+	return nil
+}
+
+// Am tries to apply the patch in the given path into the current branch
+// by performing a three-way merge (similar to git cherry-pick). It returns
+// an error if the patch cannot be applied.
+func (i *interactor) Am(path string) error {
+	i.logger.Infof("Applying patch at %s", path)
```

Comment:
IMHO this and pretty much all other logging here should IMHO be debug level, not info